seps/sep-0002.md

```diff
@@ -0,0 +1,102 @@
+# Use-variants: for packages with conflicting libraries
+
+## The problem
+
+The `intel-oneapi-mkl` package is installed with many libraries. Some
+of these libraries define the same symbols but differ in their
+implementation. For example there is a set of libraries built with
+threading support, and another set which offers the same
+functionality but are built without threading support.
+
+Right now, the only mechanism Spack offers to choose between these
+libraries is a variant: you could add a `threading` variant and then
+choose which libraries to supply to dependents based on the value
+of that variant. The problem with a variant is that different
+variant values generate distinct installations of a package: we would
+install `intel-oneapi-mkl` once for each choice of `threading`, even
+though the installations would be the same.
+
+See also: https://github.com/spack/spack/discussions/22749
+
+## Proposed Changes
+
+* Modify variant declarations with a flag to indicate that the variant
+  determines how the package is used rather than how it is installed
+* During concretization, this variant can be treated like any other
+  variant.
+* After concretization, the variant is stripped from the node and
+  stored on the edge connecting the node to its parent.
```

Comment:
I think there are three features mentioned in this SEP, but their semantics aren't really laid out here. I see:

* `interface=True`
* `request=True`
* `provider=mkl`

I think the first two should really be type=`interface`, `type=request`, etc.

The main concern I have is that the proposed changes should really include a description of how each of these options affects concretization semantics and the final DAG.

Reply:
`interface=True` has the same concretization semantics as other variants (the difference is in terms of hashing).

Likewise `provider=mkl` doesn't modify this, only restricts when the variant affects the hash.

`request=True` doesn't modify the concretization semantics or the hash: it is only theoretically useful in terms of saving time when implementing `.libs` or .headers` for a dependency but is not required (so in a sense doesn't have the same importance as `interface=True`, which is in fact required for correct semantics).

It is true that `interface=True` and `request=True` are mutually exclusive (`interface=True` is used in `.libs` and `.headers` like `request=True`, but established additional constraints), and for that reason I suppose they could be options for a variant type. Originally I found it confusing though since their use cases are different.

Reply:
> for that reason I suppose they could be options for a variant type. Originally I found it confusing though since their use cases are different.

More than that, I should say that the Boost case is simple enough that it may not need to be addressed at all here (it can definitely be addressed separately) - which I make a case for in that section.

Reply:
can you add this stuff to the SEP?

Reply:
Yes, but to be clear I think this is a reasonable way to get to that: IMO it's worth making sure we agree before I push changes to this.

Reply:
https://github.com/spack/seps/pull/2/commits/41565f3c8b9bbe5ac41d9067db68c25b4137dc8a adds some clarifications based on this conversation